site/learn/expert-advice/authentication/webauthn-explained.md

```diff
@@ -0,0 +1,478 @@
+
+Let's look at the code flow for the registration ceremony. The options are a bit more complicated than they are for the authentication ceremony. 
+
+```javascript
+options =  {
+  publicKey: {
+    rp: { id: "cosmosclownstore.com", 
+          name: "Cosmo’s Clown Store" },
+    user: { id: "1234", 
+            name: "krusty", 
+            displayName: "krusty The Clown" },
+    challenge: "...",
+    pubKeyCredParams: [ { type: "public-key", alg: -7 }]
+    //...
+  }
+}
+```
+
+You again have the `publicKey` as the top level options field, indicating that you are using WebAuthn.
+
+Then there is the `rp` object, which has information about the website that is requesting the registration operation. This includes the name to be displayed to the end user by the authenticator.
+
+You also have the `user` field, which is information used to tie the user to the credential. Additionally, to create the key, you need to specify the algorithm to use. Do this by passing the correct parameters in `pubKeyCredParams`. Above, the value is `-7`. This is an item in the [IANA COSE Algorithms registry](https://www.iana.org/assignments/cose/cose.xhtml).
+
+After creating the options object, call the `create` method on the credentials API and if it succeeds, the user will be registered with cosmosclownstore.com.
+
+```javascript
+navigator.credentials.create(options)
+  .then(function (newCredentialInfo) {
+    // Send credential info to server 
+  }).catch(function (err) {
+    // No acceptable authenticator or user refused consent. 
+  });
+```
+
+Here's an example response.
+
+```javascript
+{
+  response: {
+    clientDataJSON: { 
+      type: “webauthn.create”,
+      challenge: ... // should match initial challenge
+      origin: ... //
+    },
+    attestationObject: {
+      authData: {
+        attestedCredentialData: {
+          credentialId: ..., 
+          credentialPublicKey: ...,
+        }
+      }, // other stuff about attestation 
+    }
+  }
+}
+```
+
+The `attestationObject` contains items such as:
+
+- the credential Id 
+- the public key 
+
+Store the `credentialId` to identify the user and the public key used when verifying the assertions generated by the other ceremonies. You must also verify the `challenge` and `type` in the `clientDataJSON` object.
+
+You may have noticed that at registration, the website passes down user information. This implies that you know who the user is. Typically, you can require another form of authentication first (i.e. user and password) and then the user will associate a WebAuth credential with this account to make their future authentication easier.
+
+As the relying party (website), you may specify what type of authenticators you want to support. You can’t get too granular, but you can specify if you only want to support platform authenticators or cross-platform authenticators.
+
+Next, let's look at the attestation ceremony.
+
+## The Attestation Ceremony
```

Comment:
I haven't done any implementation for attestation yet, but it is part of the registration ceremony rather than its own thing. One of the creation options is the type of attestation the RP would like to receive (defaulting to none).

The spec (and other sources) recommend _not_ requiring attestation unless you have a strong use case for it (e.g. high-security industry or enterprise use case).

Reply:
> The spec (and other sources) recommend not requiring attestation unless you have a strong use case for it (e.g. high-security industry or enterprise use case).

Hmmm, I didn't find that in the spec.

https://www.w3.org/TR/webauthn-2/#sctn-attestation says:

"[Authenticators](https://www.w3.org/TR/webauthn-2/#authenticator) SHOULD also provide some form of [attestation](https://www.w3.org/TR/webauthn-2/#attestation), if possible."

But thanks for catching that it is part of the registration ceremony.